pkg/cluster/loadbalancerprofile.go

```diff
@@ -0,0 +1,412 @@
+			*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations = append(*lb.LoadBalancerPropertiesFormat.FrontendIPConfigurations, newFrontendIPConfig(frontendIPConfigName, frontendConfigID, obIPOrPrefix.ID))
+			outboundRuleV4FrontendIPConfig = append(outboundRuleV4FrontendIPConfig, newOutboundRuleFrontendIPConfig(frontendConfigID))
+		} else {
+			// frontendIPConfig already exists and just needs to be added to the outbound rule
+			frontendConfig := frontendIPConfigs[obIPOrPrefix.ID]
+			outboundRuleV4FrontendIPConfig = append(outboundRuleV4FrontendIPConfig, newOutboundRuleFrontendIPConfig(*frontendConfig.ID))
+		}
+	}
+
+	// update outbound-rule-v4
+	for _, outboundRule := range *lb.LoadBalancerPropertiesFormat.OutboundRules {
+		if *outboundRule.Name == outboundRuleV4 {
+			outboundRule.OutboundRulePropertiesFormat.FrontendIPConfigurations = &outboundRuleV4FrontendIPConfig
+			break
+		}
+	}
+}
+
+// Delete all managed outbound IPs that are not in use by the load balancer.
+// The default outbound ip is saved if the api server is public.
+func (m *manager) deleteUnusedManagedIPs(ctx context.Context) error {
+	resourceGroupName := stringutils.LastTokenByte(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID, '/')
+	infraID := m.doc.OpenShiftCluster.Properties.InfraID
+
+	managedIPs, err := m.getClusterManagedIPs(ctx)
+	if err != nil {
+		return err
+	}
+
+	lb, err := m.loadBalancers.Get(ctx, resourceGroupName, infraID, "")
+	if err != nil {
+		return err
+	}
+
+	outboundIPs := getOutboundIPsFromLB(lb)
+	outboundIPMap := make(map[string]api.ResourceReference, len(outboundIPs))
+	for i := 0; i < len(outboundIPs); i++ {
+		outboundIPMap[strings.ToLower(outboundIPs[i].ID)] = outboundIPs[i]
+	}
+	var cleanupErrors []string
+	for _, ip := range managedIPs {
+		// don't delete api server ip
+		if *ip.Name == infraID+"-pip-v4" && m.doc.OpenShiftCluster.Properties.APIServerProfile.Visibility == api.VisibilityPublic {
+			continue
+		}
+		if _, ok := outboundIPMap[strings.ToLower(*ip.ID)]; !ok && strings.Contains(strings.ToLower(*ip.ID), strings.ToLower(m.doc.OpenShiftCluster.Properties.ClusterProfile.ResourceGroupID)) {
+			ipName := stringutils.LastTokenByte(*ip.ID, '/')
+			m.log.Infof("deleting managed public IP Address: %s", ipName)
+			err := m.publicIPAddresses.DeleteAndWait(ctx, resourceGroupName, ipName)
+			if err != nil {
+				cleanupErrors = append(cleanupErrors, fmt.Sprintf("deletion of unused managed ip %s failed with error: %v", ipName, err))
+			}
+		}
+	}
+	if cleanupErrors != nil {
+		return fmt.Errorf("failed to cleanup unused managed ips\n%s", strings.Join(cleanupErrors, "\n"))
+	}
+
+	return nil
+}
+
+// Returns the desired RP managed outbound publicIPAddresses.  Additional Managed Outbound IPs
+// will be created as required to satisfy ManagedOutboundIP.Count.
+func (m *manager) getDesiredOutboundIPs(ctx context.Context) ([]api.ResourceReference, error) {
```

Comment:
nit: similar to earlier - this `get` function is mutating since reconcileDesiredManagedIPs does things like create new IP addresses.

Reply:
I will address this in a follow up PR.